File: moxy/src/main/java/com/omegar/mvp/viewstate/ViewCommands.java
package com.omegar.mvp.viewstate;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import com.omegar.mvp.MoxyReflector;
import com.omegar.mvp.MvpView;
import com.omegar.mvp.viewstate.strategy.StateStrategy;

/**
 * Date: 17.12.2015
 * Time: 11:09
 *
 * @author Yuri Shmakov
 */
public class ViewCommands<View extends MvpView> {
	private final List<ViewCommand<View>> mState = new ArrayList<>();

	public void beforeApply(ViewCommand<View> viewCommand) {
		StateStrategy stateStrategy = getStateStrategy(viewCommand);

		stateStrategy.beforeApply(mState, viewCommand);
	}

	public void afterApply(ViewCommand<View> viewCommand) {
		StateStrategy stateStrategy = getStateStrategy(viewCommand);

		stateStrategy.afterApply(mState, viewCommand);
	}

	private StateStrategy getStateStrategy(ViewCommand<View> viewCommand) {
		return (StateStrategy) MoxyReflector.getStrategy(viewCommand.getStrategyClass());
	}

	public boolean isEmpty() {
		return mState.isEmpty();
	}

	public void reapply(View view, Set<ViewCommand<View>> currentState) {
		final ArrayList<ViewCommand<View>> commands = new ArrayList<>(mState);

		for (ViewCommand<View> command : commands) {
			if (currentState.contains(command)) {
				continue;
			}

			command.apply(view);

			afterApply(command);
		}
	}

	public List<ViewCommand<View>> getCurrentState() {
		return mState;
	}
}

File: moxy/src/main/java/com/omegar/mvp/viewstate/ViewCommands.kt
package com.omegar.mvp.viewstate

import com.omegar.mvp.MvpView

/**
 * Date: 17.12.2015
 * Time: 11:09
 *
 * @author Yuri Shmakov
 */
@JvmInline
value class ViewCommands<View : MvpView>(val list: MutableList<ViewCommand<View>> = ArrayList()) {

    private companion object {
        private const val KEY_STATE = "state"
    }

    fun beforeApply(command: ViewCommand<View>) = command.beforeApply(list)

    fun afterApply(command: ViewCommand<View>) = command.afterApply(list)

    @Suppress("UNCHECKED_CAST")
    fun <C : ViewCommand<View>> findCommand(clz: Class<*>): C? = list.lastOrNull { it.javaClass === clz } as? C?

    fun reapply(view: View, currentState: Set<ViewCommand<View>>) {
        list.filter { it !in currentState }
                .forEach { command ->
                    command.apply(view)
                    afterApply(command)
                }
    }

    override fun toString(): String {
        return "ViewCommands{" +
                "state=" + list +
                '}'
    }

}